File: git_extract.rb
require 'set'
require 'rubygems'
require 'sequel'
require 'open4'
require 'time'
require 'trollop'

class String 
	def to_proc
		proc { |*args| send(self, *args) }
	end
end

class Symbol
	def to_proc
		proc { |obj, *args| obj.send(self, *args) }
	end
end

def setup_table(name, db, reset, &block)
	db.drop_table name if db.table_exists? name and reset
	if not db.table_exists? name
		db.create_table name, &block 
	end
end

class Timer
	def initialize
		@last = Time.now
	end
	
	def last
		@last
	end

	def delta
		last = @last
		@last = Time.now
		return @last - last
	end

	def log
		puts delta
	end
end

$timer = Timer.new


def setup_tables(db, reset)
	setup_table :git_repo, db, reset do
		String :commit 
		String :repo
		primary_key( :commit, :repo)
		index :commit
	end
	setup_table :git_commit, db, reset do
		String :commit
		String :tree
		String :author
		DateTime :author_dt
		String :author_id
		String :committer
		DateTime :committer_dt
		String :committer_id
		String :subject, :default => ''
		Integer :num_children, :default => 0
		Integer :num_parents, :default => 0
		String :log, :default => ''
		primary_key(:commit)
	end

	setup_table :git_dag, db, reset do
		String :child, :null => false
		String :parent, :null => false			
		index :parent
		index :child
	end
	setup_table :git_revision, db, reset do	
		String :commit, :null => false
		Integer :add
		Integer :remove
		String :path, :null => false
		primary_key(:commit, :path)
		index(:commit)
		index(:path)
	end

	setup_table :git_refs_tags, db, reset do
		String :commit, :null => false
		String :path, :null => false
		primary_key(:commit, :path)
	end

	setup_table :git_chain, db, reset do
		String :commit, :null => false
		String :name_addr 
		Integer :indiv_id
		String :type # s = Signed, a = Acked, t = Tested, r = Reviewed, c = Cc'ed
		primary_key :commit
		index :commit
	end

end

$git_log_cmd = "git log --full-history --all --numstat -M -C --pretty=format:\"" +
		"__START_GIT_COMMIT_LOG_MSG__%n%H%n%T%n%an <%ae>%n%aD%n%cn <%ce>" +
		"%n%cD%n%P%n%d%n%s%n%b%n__END_GIT_COMMIT_LOG_MSG__\""

def get_git_log_lines(repo_path)
	#run the log command
	git_log_cmd = "cd #{repo_path} && #{$git_log_cmd}"
	puts "Getting git log text"
	puts git_log_cmd
	error_txt = ""
	history = []
	status = Open4::popen4(git_log_cmd) do |pid, stdin, stdout, stderr|
		history = stdout.readlines
		error_txt = stderr.read
	end
	#if there was an error message, then print it out and die
	if status.exitstatus != 0
		puts "exit status is : #{status.exitstatus}"	
		puts error_txt
		exit(1)
	end

	puts "Parsing git log"
	
	return history.map!( &:strip )
end

def get_file_log_lines(file)
	return open(file).readlines.map( &:strip )
end


def parse_log(repo_name, history, db)
	print "**** parsing history ****\n"

	i = 0
	parsed_hashes = 0

	#get all commits that are already in the db
	commits = Set.new
	db[:git_commit].each { |commit| commits.add(commit) }

	git_commit_ps = db[:git_commit].prepare(:insert, :commit => :$commit, :tree => :$tree, 
		:author => :$author, :author_dt => :$author_dt,
		:committer => :$committer, :committer_dt => :$committer_dt,
		:log => :$log)

	history_length = history.length

	commit_data = []
	repo_data = []
	refs_tags_data = []
	dag_data = []
	revision_data = []
	db.transaction do	
	loop do
		i += 1 while i < history.length and not history[i] =~ /^__START_GIT_COMMIT_LOG_MSG__/
		break if i >= history_length
		i += 1
		s = i
		commit_id = history[i]

		#we ALWAYS have to insert the repo and the refs
		repo_data << [repo_name, commit_id]

		i += 1
		refs_line = history[s+7]
		if not refs_line.empty?
			refs_line.gsub!(/^\s*\(/, "")
			refs_line.gsub!(/\)\s*$/, "")
			#continue from here
			refs_line.split(", ").each { |ref| 
				refs_tags_data << [commit_id, ref]
			}
		end

		parsed_hashes += 1
		if parsed_hashes % 1000 == 0
			puts "parsed #{parsed_hashes} commits\n" 
			puts "this took #{$timer.delta} seconds\n"
		end
		#the sha may already be in the database so check that
		next if commits.include? commit_id

		log = ""
		i = s + 8
		while i < history.length and not history[i] =~ /^__END_GIT_COMMIT_LOG_MSG__/
			log << history[i] + "\\n"
			i += 1
		end
		tree = history[s+1]
		author = history[s+2]
		author_dt = history[s+3]
		committer = history[s+4]
		committer_dt = history[s+5]
		commit_data << [commit_id, tree, author, author_dt, committer, 
			committer_dt, log]
		# line 6 contains the parents, add those
		history[s+6].split(/\s+/).each do |parent| 
			dag_data << [commit_id, parent]
		end

		#move past the __END_GIT_COMMIT_LOG_MSG__
		i += 1
		# lines after this are the files that were changed and counts of lines changed
		while i < history_length and not history[i] =~ /__START_GIT_COMMIT_LOG_MSG__/
			if history[i] =~ /^(\d+)\s+(\d+)\s+(.*)/
				# (lines added)\s+(lines removed)\s+(path)
				revision_data << [commit_id, $1.to_i, $2.to_i, $3]
			elsif history[i] =~ /^\-\s+\-\s+(.*)/
				revision_data << [commit_id, nil, nil, $1]
			end
			i += 1
		end
	end
	end
	history.clear
	$timer.log
	puts "inserting #{revision_data.length} revision records"
	db[:git_revision].multi_insert([:commit, :add, :remove, :path], revision_data)
	$timer.log
	puts "inserting #{dag_data.length} dag records"
	db[:git_dag].multi_insert([:child, :parent], dag_data)
	$timer.log
	puts "inserting #{commit_data.length} commit records"
	db[:git_commit].multi_insert( [:commit, :tree, :author, :author_id,
		:committer, :committer_dt, :log], commit_data)
	$timer.log
	puts "inserting #{refs_tags_data.length} ref/tag records"
	db[:git_refs_tags].multi_insert([:commit, :path], refs_tags_data)
	$timer.log
	puts "inserting #{repo_data.length} repo records"
	db[:git_repo].multi_insert([:repo, :commit], repo_data)
	$timer.log
end	

#do some post processing and fill in the tables
def update_relations(db)
	puts "gathering number of parents"
	db << "update git_commit set num_parents = r.parents from (select child, count(*) as parents from
		git_dag group by child) as r where r.child = git_commit.commit"
	$timer.log
	puts "gathering number of children"
	db << "update git_commit set num_children = r.children from (select parent, count(*) as children from
		git_dag group by parent) as r where r.parent = git_commit.commit"
	$timer.log
end




def main
	opts = Trollop::options do
		version "git_extract.rb version 0.1"
		banner <<-EOS
git_extract.rb will mine information from git and put it into
a database. You must specify ONE of: repo location with -r or
log file with -l.  The log file must be the result of running
(all on one line, no line breaks):

#{$git_log_cmd}
EOS
		opt :repo, "path to git repository", :short => "r", :type => String 
		opt :log, "log file for git repository", :short => "l", :type => String
		conflicts :log, :repo
		opt :dburl, "url for database example: postgres://cabird:passwd@localhost/git_db", 
			:short => "d", :required => true, :type => String
		opt :name, "repo name to be stored in database", :short => "n", :required => true,
			:type => String
	end
	p opts
	db = Sequel.connect(opts[:dburl])
	setup_tables(db, true)
	puts "getting log lines"
	if opts[:log] != nil
		history = get_file_log_lines(opts[:log])
	elsif opts[:repo] != nil
		history = get_git_log_lines(opts[:repo])
	else
		puts "error! no way to get log lines, must specify either location of repo or a log file"
		exit(1)
	end
	$timer.log
	parse_log(opts[:name], history, db)
	update_relations(db)
end

main() if __FILE__ == $0

File: git_extract.pl
#!/usr/bin/perl -w
#test -- my @lines = split(/\n/, qx(git log --since="four weeks ago" --pretty=format:"%H %T %an <%ae> %aD %cn <%ce>% %cD %P%n") );

use warnings;
use strict;

use DBI;
use Cwd;

#http://www.perlmonks.org/?node_id=591180
use utf8; 
use Encode;

#stop stdout buffering
$| = 1;

sub process_git_repo ($$) {
	my ($db_name, $path) = @_;

	my $dbh = DBI->connect("dbi:Pg:database=$db_name", '', '', {AutoCommit => 1});

	my $check_sha = $dbh->prepare(q{select commit from git_commit where commit = ?});

	#This next line helps decipher the git command line and the regexp
	my $insert_git_commit = $dbh->prepare(q{insert into git_commit(commit, tree, author, author_dt, committer, committer_dt, subject, log) values(?,?,?,?,?,?,?,?)});

	my $insert_dag = $dbh->prepare(q{insert into git_dag(child, parent) values (?,?)});

	my $insert_revisions = $dbh->prepare(q{insert into git_revision(commit, add, remove, path) values (?,?,?,?)});

	my $insert_refs_tags = $dbh->prepare(q{insert into git_refs_tags(commit, path) values (?,?)});

	my $insert_git_repo = $dbh->prepare(q{insert into git_repo(commit, repo) values(?,?)});

	my @history = split(/\n/, qx(git log --full-history --all --numstat -M -C --pretty=format:"-prxh-START_LOG_MSG-%n%H%n%T%n%an <%ae>%n%aD%n%cn <%ce>%n%cD%n%P%n%d%n%s%n%b%n-prxh-END_LOG_MSG-%n")); 
	
	if (!defined $history[0]) {
		print STDERR "Error processing git repo $path\n";
		return;
	}

	#remove anything before first commit sha1
	while (@history) {
		my $line = shift @history;
		if ($line =~ /^-prxh-START_LOG_MSG-$/) {
			last;
		}
	}

	LOG:
	#if there isn't at least 9 lines then we've run out of history
	for (my $i = 0; $i < @history - 9 ; $i ++) {

		#we will do the insert after we get the log so record the offset in $s (or start)
		my $s = $i;
		my $commit_id = $history[$i];
		
		#test -- print "$history[$i]\n$history[$i+1]\n$history[$i+2]\n$history[$i+3]\n$history[$i+4]\n";
		#	 print "$history[$i+5]\n$history[$i+6]\n$history[$i+7]\n$history[$i+8]\n";


		#
		#We always have to insert the repo and refs, tags, etc
		$insert_git_repo->execute($commit_id, encode("UTF-8", $path)) or warn "Problem inserting repo name $commit_id ", $dbh->errstr;

		#refs, tags, etc (%d in git-log V 1.6.? or higher)
		if(defined $history[$s+7]) {
			$history[$s+7] =~ s/^\s*\(//;	
			$history[$s+7] =~ s/\)\s*$//;	
			foreach my $ref (split(/, /, $history[$s+7])) {
				#print "$path$ref\n";
				$insert_refs_tags->execute($commit_id, encode("UTF-8", "$path$ref")) 
					or warn "Problem inserting refs/tags on $commit_id ", $dbh->errstr;
			}
		}

		#
		#Is the sha already in the db?
		$check_sha->execute($commit_id) or warn "Problem checking for existing commit sha $commit_id ", $dbh->errstr;
		#if yes, move to next log
		if ($check_sha->rows > 0) {
			$i += 9;
			while ($i < @history) {
				if ($history[$i] =~ /^-prxh-START_LOG_MSG-$/) {
					next LOG;
				}
				$i ++;
			}
			#if we are here, we have finished the file
			last LOG;
		}

		#
		#We need to get the log before we insert the info about the commit
		#
		#the subject line sometimes ends in \n and sometimes doesn't, 
		# so we put it before log and remove any trailing or leading NEW_LINE from the log
		#
		$i = $s + 9; #The log is variable length so move $i
		my $log = '';
		while($i < @history && $history[$i] !~ /^-prxh-END_LOG_MSG-/) {
			$log .= "$history[$i]NEW_LINE";
			$i ++;	
		}
		$i ++; #move past -prxh-END_LOG_MSG-
		#remove extra trailing or leading NEW_LINE
		$log =~ s/^\s*NEW_LINE//;
		$log =~ s/NEW_LINENEW_LINE\s*$/NEW_LINE/;
		#print "$log\n";	

		#now do the insert
		$insert_git_commit->execute($history[$s], $history[$s+1], encode("UTF-8", $history[$s+2]), $history[$s+3], 
			encode("UTF-8", $history[$s+4]), $history[$s+5], 
			encode("UTF-8", $history[$s+8]), encode("UTF-8", $log))
				or warn "Problem inserting commit info on $commit_id ", $dbh->errstr;

		#	
		#create the dag
		#parents (%P in git-log)
		if (defined $history[$s+6]) {
			foreach my $parent (split(/ /, $history[$s+6])) {
				#print "$commit_id -- $parent\n";
				$insert_dag->execute($commit_id, $parent) or warn "Problem inserting parent on $commit_id \nThere was an error in git where the parent might be inserted more than once on some merges, this will exist in certain old histories, so check the log on this commit!\n", $dbh->errstr;
			}
		}	

		#file revisions
		while($i < @history && $history[$i] !~ /^-prxh-START_LOG_MSG-$/) {
			if ($history[$i] =~ /^(\d+)\s* (\d+)\s* (.*)/xms) {
				#print "adds = $1 -- removes $2 -- path = $3\n";
				$insert_revisions->execute($commit_id, $1, $2, $3) 
					or warn "Problem inserting file history on $commit_id ", $dbh->errstr;
			}
			elsif ($history[$i] =~ /^\-\s* \-\s* (.*)/xms) {
				#print "binary file path = $1\n";
				$insert_revisions->execute($commit_id, undef, undef, $1)
					or warn "Problem inserting binary file history on $commit_id ", $dbh->errstr;
			}

			$i ++;
		}
		#print "\n\n";
	}

	$check_sha->finish;
	$insert_git_repo->finish;
	$insert_refs_tags->finish;
	$insert_revisions->finish;
	$insert_dag->finish;
	$insert_git_commit->finish;
	$dbh->disconnect;

}

#
# Step through all the directories
#
sub process_dir 
{
	my ($path, $config_ref, $db_check_repo_ref) = @_;

	chdir $path or die "Cannot processes $path";
	my $cwd = getcwd;

	if (-e ".git") 
	#if ($path =~ /\w+\.git\s*$/xms)
	{
		if ($cwd =~ /$config_ref->{repo_path_ignore}(.*)/) {
			my $short_path = $1;
			$db_check_repo_ref->execute($short_path) or warn "Problem checking if repo has already been processed\n";
			if ($db_check_repo_ref->rows > 0) {
				print "Skipping ... we've already done this repo: $short_path\n";
			}
			else {
				print "Processing git repo $short_path \n";
				process_git_repo($config_ref->{db_name}, $short_path);
			}
		}
		else {
			warn "Problem processing cwd: $cwd";
		}
	}
	else {
		foreach my $next_path (<*>) 
		{
			if (-d $next_path) 
			{
				process_dir($next_path, $config_ref, $db_check_repo_ref);
			}
		}
	}
	chdir '..' or die "Cannot backout of $path";
}

use Config::General;

my $config_path = shift @ARGV;
if (!defined $config_path) {
	$config_path = 'config';
}
die "Config file \'$config_path\' does not exist"
	unless (-e $config_path);
my %config =  Config::General::ParseConfig($config_path);
#test -- print "$config{repo_path}, $config{db_name}\n";

my $dbh_ref = DBI->connect("dbi:Pg:database=$config{db_name}", '', '', {AutoCommit => 1});

my $db_check_repo_ref = $dbh_ref->prepare(q{select repo from git_repo where repo = ? group by repo}); 

process_dir($config{repo_path}, \%config, $db_check_repo_ref);

#update parent and children refs in git table
$dbh_ref->do(q{update git_commit set num_parents = r.parent from (select child, count(*) as parent from git_dag, git_commit where git_dag.child = git_commit.commit group by child) as r where r.child = git_commit.commit})
	or warn "Problem updating number of parents ", $dbh_ref->errstr;

$dbh_ref->do(q{update git_commit set num_children = r.child from (select parent, count(*) as child from git_dag, git_commit where git_dag.parent = git_commit.commit group by parent) as r where r.parent = git_commit.commit})
	or warn "Problem updating number of children ", $dbh_ref->errstr;

$db_check_repo_ref->finish;
$dbh_ref->disconnect;
 
__END__

=head1 Extracts a git repo from the logs

make sure you run git_extract.sql first

You need to set up a config file with the following
db_name=linux3
repo_path=/home/pcr/mining_procedure/test/git_repo_test
repo_path_ignore=/home/pcr/mining_procedure/

=cut
